presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java

```diff
@@ -196,111 +199,118 @@ public Page getOutput()
-        int maxEntries = getCurrentMaxEntries();
-
-        // Append elements repeatedly to replicate output columns
-        replicatedBlockBuilders.forEach(blockBuilder -> blockBuilder.appendRepeated(currentPosition, maxEntries));
-
-        // Process this position in unnesters
-        unnesters.forEach(unnester -> unnester.processCurrentAndAdvance(maxEntries));
+        if (nestedType instanceof ArrayType) {
+            Type elementType = ((ArrayType) nestedType).getElementType();
 
-        if (withOrdinality) {
-            for (long ordinalityCount = 1; ordinalityCount <= maxEntries; ordinalityCount++) {
-                BIGINT.writeLong(ordinalityBlockBuilder, ordinalityCount);
+            if (!isLegacyUnnest && elementType instanceof RowType) {
+                Type[] types = Iterables.toArray(requireNonNull(elementType, "elementType is null").getTypeParameters(), Type.class);
+                return new ArrayOfRowsUnnester(types);
+            }
+            else {
+                return new ArrayUnnester(elementType);
             }
         }
-
-        return maxEntries;
+        else if (nestedType instanceof MapType) {
+            Type keyType = ((MapType) nestedType).getKeyType();
+            Type valueType = ((MapType) nestedType).getValueType();
+            return new MapUnnester(keyType, valueType);
+        }
+        else {
+            throw new IllegalArgumentException("Cannot unnest type: " + nestedType);
+        }
     }
 
-    private int getCurrentMaxEntries()
+    private void resetBlockBuilders()
     {
-        return unnesters.stream()
-                .mapToInt(Unnester::getCurrentUnnestedLength)
-                .max()
-                .orElse(0);
+        for (int i = 0; i < replicateTypes.size(); i++) {
+            Block newInputBlock = currentPage.getBlock(replicateChannels.get(i));
+            replicatedBlockBuilders.get(i).resetInputBlock(newInputBlock);
+        }
+
+        for (int i = 0; i < unnestTypes.size(); i++) {
+            int inputChannel = unnestChannels.get(i);
+            Block unnestChannelInputBlock = currentPage.getBlock(inputChannel);
+            unnesters.get(i).resetInput(unnestChannelInputBlock);
+        }
     }
 
-    private void prepareForNewOutput(PageBuilderStatus pageBuilderStatus)
+    private void calculateMaxEntries()
     {
-        unnesters.forEach(unnester -> unnester.startNewOutput(pageBuilderStatus, estimatedMaxRowsPerBlock));
-        replicatedBlockBuilders.forEach(replicatedBlockBuilder -> replicatedBlockBuilder.startNewOutput(estimatedMaxRowsPerBlock));
+        int positionCount = currentPage.getPositionCount();
+        maxEntries = ensureCapacity(maxEntries, positionCount, SMALL, INITIALIZE);
+
+        for (Unnester unnester : unnesters) {
+            unnester.populateEntryCounts(positionCount);
+            for (int i = 0; i < positionCount; i++) {
+                int curLen = unnester.getEntryCounts()[i];
```

Comment:
don't abbreviate; perhaps, change Unnested.getLengths() method to `getLength(int position)` and write this like so:

```
maxEntries[i] = max(maxEntries[i], unnester.getLength(i));

    public int getLength(int position)
    {
        return lengths[position];
    }
```

Reply:
getEntryLengths() is still required, therefore, I added the function getEntryLength(int position) as follows:

```
public int[] getEntryLengths()
    {
        return entryLengths;
    }
    
    public int getEntryLength(int position)
    {
        assert position >= 0 && position < entryLengths.length;
        return entryLengths[position];
    }
```